extensions/positron-notebook-controllers/src/notebookSessionService.ts

```diff
@@ -87,6 +87,31 @@ export class NotebookSessionService implements vscode.Disposable {
 		return this._notebookSessionsByNotebookUri.has(notebookUri);
 	}
 
+	/**
+	 * Checks for a restarting notebook for the given notebook URI.
+	 *
+	 * @param notebookUri The notebook URI to check for.
+	 * @returns True if a session is restarting for the given notebook URI.
+	 */
+	hasRestartingNotebookSession(notebookUri: vscode.Uri): boolean {
+		return this._restartingSessionsByNotebookUri.has(notebookUri);
+	}
+
+	/**
+	 * Wait for a notebook session to complete a restart sequence.
+	 *
+	 * @param notebookUri The notebook URI to wait for.
+	 * @returns A promise that resolves when the session has completed the restart sequence.
+	 * @throws If no session is restarting for the given notebook URI.
+	 */
+	async waitForNotebookSessionToRestart(notebookUri: vscode.Uri): Promise<positron.LanguageRuntimeSession> {
+		const restartingSessionPromise = this._restartingSessionsByNotebookUri.get(notebookUri);
+		if (!restartingSessionPromise) {
+			throw new Error(`No session is restarting for notebook ${notebookUri.path}`);
+		}
+		return restartingSessionPromise;
+	}
```

Comment:
If you don't ever see `hasRestartingNotebookSession` outside of this usecase I almost feel like consolidating to something like `waitIfNotebookSessionIsRestarting` that does the check you do in `notebookController.ts` would simplify things. E.g. you always await but if there's not a restarting notebook session it just immediately returns.